lang/cs/Org.Apache.REEF.Network/Elastic/Driver/Default/DefaultElasticTaskSetManager.cs

```diff
@@ -0,0 +1,1344 @@
+                            {
+                                Log.Log(Level.Warning, msg + " Retry");
+                                System.Threading.Tasks.Task.Run(() =>
+                                {
+                                    Thread.Sleep(_parameters.WaitTime);
+                                    SendToTasks(new List<IElasticDriverMessage>() { returnMessage }, retry + 1);
+                                });
+                            }
+                            else if (retry >= _parameters.Retry)
+                            {
+                                Log.Log(Level.Warning, msg + " Aborting");
+                                Fail(returnMessage.Destination);
+                            }
+                            else
+                            {
+                                Log.Log(Level.Warning, msg + " Ignoring");
+                            }
+
+                            continue;
+                        }
+
+                        taskInfo.TaskRunner.Send(returnMessage.Serialize());
+                    }
+                }
+            }
+        }
+
+        private void SpawnNewEvaluator(int id)
+        {
+            Log.Log(Level.Warning, "Spawning new evaluator for id {0}", id);
+
+            var request = _evaluatorRequestor.NewBuilder()
+                .SetNumber(1)
+                .SetMegabytes(_parameters.NewEvaluatorMemorySize)
+                .SetCores(_parameters.NewEvaluatorNumCores)
+                .SetRackName(_parameters.NewEvaluatorRackName)
+                .Build();
+
+            _evaluatorRequestor.Submit(request);
+        }
+
+        private void Reschedule(RescheduleEvent rescheduleEvent)
+        {
+            var id = Utils.GetTaskNum(rescheduleEvent.TaskId) - 1;
+
+            lock (_taskInfos[id])
+            {
+                _taskInfos[id].NumRetry++;
+
+                if (_taskInfos[id].NumRetry > _parameters.NumTaskFailures)
+                {
+                    Log.Log(Level.Error, "Task {0} failed more than {1} times: aborting",
+                        rescheduleEvent.TaskId,
+                        _parameters.NumTaskFailures);
+                    Fail(rescheduleEvent.TaskId);
+                }
+
+                if (rescheduleEvent.Reschedule)
+                {
+                    Log.Log(Level.Info, "Rescheduling task {0}", rescheduleEvent.TaskId);
+
+                    _taskInfos[id].RescheduleConfigurations = rescheduleEvent.RescheduleTaskConfigurations;
+
+                    SubmitTask(id);
+                }
+            }
+        }
+
+        private void Fail(string taskId = "")
+        {
+            IFailureEvent @event = new FailEvent(taskId);
+
+            EventDispatcher(ref @event);
+        }
+
+        private void LogFinalStatistics()
+        {
+            Log.Log(Level.Info, "Total Failed Tasks: {0}\nTotal Failed Evaluators: {1}\n{2}",
```

Comment:
wrap into
```C#
if (Log.IsLoggable(Level.Info))
```
to avoid calling `string.Join()` when not necessary

Reply:
Very good point! Fixed.